template/template.go

```diff
@@ -1,46 +1,136 @@
 package template
 
 import (
+	"strings"
+
 	"github.com/hashicorp/hil"
 	"github.com/hashicorp/hil/ast"
+	"github.com/openpixel/rise/config"
 	"github.com/openpixel/rise/interpolation"
 )
 
 // Template is a container for holding onto the ast Variables
 type Template struct {
-	vars map[string]ast.Variable
+	vars      map[string]ast.Variable
+	templates map[string]ast.Variable
 }
 
 // NewTemplate will prepare a template object for use
-func NewTemplate(vars map[string]ast.Variable) (*Template, error) {
-	if vars == nil {
-		vars = make(map[string]ast.Variable)
-	}
+func NewTemplate(configResult *config.Result) (*Template, error) {
 	return &Template{
-		vars: vars,
+		vars:      configResult.Variables,
+		templates: configResult.Templates,
 	}, nil
 }
 
 func (t *Template) buildConfig() *hil.EvalConfig {
+	vars := make(map[string]ast.Variable)
+
+	for k, v := range t.vars {
+		vars[k] = v
+	}
+	for k, v := range t.templates {
+		vars[k] = v
+	}
+
 	return &hil.EvalConfig{
 		GlobalScope: &ast.BasicScope{
 			FuncMap: interpolation.CoreFunctions,
-			VarMap:  t.vars,
+			VarMap:  vars,
 		},
 	}
 }
 
 // Render will parse the provided text and interpolate the known variables/functions
 func (t *Template) Render(text string) (hil.EvaluationResult, error) {
+	var resultErr error
+
+	config := t.buildConfig()
+
 	tree, err := hil.Parse(text)
 	if err != nil {
 		return hil.InvalidResult, err
 	}
 
-	result, err := hil.Eval(tree, t.buildConfig())
+	vf := visitorFn{
+		config:    config,
+		templates: t.templates,
+	}
+	tree = tree.Accept(vf.fn)
+	if vf.resultErr != nil {
+		return hil.InvalidResult, resultErr
+	}
+
+	result, err := hil.Eval(tree, config)
 	if err != nil {
 		return hil.InvalidResult, err
 	}
 
 	return result, nil
 }
+
+type visitorFn struct {
+	resultErr error
+
+	config *hil.EvalConfig
+
+	templates map[string]ast.Variable
+}
+
+func (vf visitorFn) fn(n ast.Node) ast.Node {
+	if vf.resultErr != nil {
+		return n
+	}
+
+	var varName string
+	switch vn := n.(type) {
+	case *ast.VariableAccess:
+		varName = vn.Name
+	case *ast.Index:
+		if va, ok := vn.Target.(*ast.VariableAccess); ok {
```

Comment:
I actually need to run the template function whenever a reference to `*ast.VariableAccess` is found. at the moment, if it finds a Target & Key, it will only run it once.

On top of that, when it finds an `*ast.Index`, that is when I need to look at adding a "index" variable to the substitution.